Migrated to Hive

schwartz-concordium opened the pull request:
## Purpose

Migrating from Shared Preferences to Hive.
https://concordium.atlassian.net/browse/CBW-1474

## Changes

* Migrated from Shared Preferences to Hive
* Renamed from `<Prefix>Service` to `<Prefix>Provider` for `StorageProvider` to follow the naming conventions of the Bloc architecture.
https://bloclibrary.dev/#/architecture?id=data-provider
* Renamed pipeline to just `ci.yml` since test step has been added.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.

bisgardo committed:
Improve initialization and network structure

Commit:
migrated to Hive

Commit:
Merge branch 'main' into cbw-1474/migrate-to-hive

Commit:
changed name

Commit:
added test

Commit:
updated ci to include test

Commit:
added test location and updated gitignore

Commit:
adding test for all cases

Commit:
Merge branch 'main' into cbw-1474/migrate-to-hive

Commit:
Fix merge errors

Commit:
Update ci.yml

Commit:
formatted

orhoj commented on pubspec.yaml:
I believe that something had to be done manually when updating dependencies for them to propagate to the iOS project. But I don't see any changes to the iOS files in this PR.

bisgardo replied:
Nice catch! I've pushed a commit with the changes from running the branch on my mac.

Note that it also caused some changes to `pubspec.yaml` that look a little weird to me. Unsure if it's caused by me upgrading the flutter tool locally. @schwartz-concordium

schwartz-concordium replied:
@bisgardo even after running `flutter upgrade` and `flutter pub get` and can’t get the `pubspec.lock` file to look like what has been committed. The versions seem downgraded.

orhoj commented on lib/providers/storage.dart:
```diff
@@ -0,0 +1,44 @@
+  }
+
+  /// Reads the currently accepted T&C version.
+  AcceptedTermsAndConditions? getAcceptedTermsAndConditions(NetworkName networkName) {
```
The terms and conditions are universal and not per network. You only accept them once for a single installation.

bisgardo replied:
Yes and we should probably also think of a better way to work with the valid T&C version than fetching it in a network-specific context.

orhoj commented on lib/providers/storage.dart:
```diff
@@ -0,0 +1,44 @@
+    _registerAdapters();
+    await _openBoxes();
+
+    return StorageProvider._(await Hive.openBox<AcceptedTermsAndConditions>(AcceptedTermsAndConditions.table));
```
It might be because I don't know Hive well (yet), but you also open the box in `_openBoxes()`. So to me it appears like we do that twice which is weird.

orhoj replied:
So I just checked the documentation and it states that:
```
If the box is already open, the instance is returned and all provided parameters are being ignored.
```
But for clarity we should use:
```
    return StorageProvider._(Hive.box<AcceptedTermsAndConditions>(AcceptedTermsAndConditions.table));
```
instead.

schwartz-concordium replied:
The idea was to prepare for more boxes to be open asynchronously in `_openBoxes`. I didn’t want to change to return type every time a new box was added.

I changed to your fix by changing the input to the constructor.

orhoj commented on lib/entities/accepted_terms_and_conditions.dart:
```diff
@@ -0,0 +1,26 @@
+  static const table = "accepted_terms_and_conditions";
+
+  @HiveField(1)
+  final String acceptedVersion;
```
As the object is already named `AcceptedTermsAndConditions`, then to me it is excessive to also call the variable `acceptedVersion` instead of just `version`.

orhoj commented on lib/state/services.dart:
```diff
@@ -36,10 +36,10 @@ class ServiceRepository {
-  /// Global service for interacting with shared preferences.
-  final SharedPreferencesService sharedPreferences;
+  /// Global service for interacting with storage.
+  final StorageProvider storage;
```
We don't need it yet (as the terms and conditions acceptance is global), but I think we should place the `StorageProvider` as a network service as most interactions will be parameterized by the currently active network. Just something to keep in mind, but not something that must be done in this PR.

bisgardo replied:
I would strongly prefer a single datastore with the network etc. being a field that we can filter on. We can add network-specific facade that does this filtering (unsure it that's what you meant @orhoj).

schwartz-concordium replied:
As described below refactoring has been done into using a single storage provider and then possible network filterings done in repositories.

orhoj commented on lib/state/terms_and_conditions.dart:
```diff
@@ -101,17 +91,17 @@ class TermsAndConditionAcceptance extends Cubit<TermsAndConditionsAcceptanceStat
-    if (nextAcceptedVersion == currentAcceptedVersion) {
-      return Future.value();
-    }
+  Future<void> _persistAcceptedVersionIfChanged(AcceptedTermsAndConditions? nextAcceptedVersion) {
```
The method name no longer reflects what this does. This always persists something.

orhoj commented on lib/state/terms_and_conditions.dart:
```diff
@@ -101,17 +91,17 @@ class TermsAndConditionAcceptance extends Cubit<TermsAndConditionsAcceptanceStat
   void onChange(Change<TermsAndConditionsAcceptanceState> change) {
     super.onChange(change);
 
+    if (change.currentState == change.nextState) {
```
You can simplify this by negating the expression:
```
if (change.currentState != change.nextState) {
  _persistAcceptedVersionIfChanged(change.nextState.accepted);
}
```

orhoj commented on lib/state/terms_and_conditions.dart:
```diff
@@ -49,12 +38,13 @@ class TermsAndConditionsAcceptanceState {
 class TermsAndConditionAcceptance extends Cubit<TermsAndConditionsAcceptanceState> {
   /// Service used to persist the accepted T&C version.
-  final SharedPreferencesService _prefs;
+  final StorageProvider _storage;
```
If we go by the naming convention used by block, then I think you should instead create a `TermsAndConditionsRepository` that wraps the `StorageProvider` to expose the terms and conditions specific functionality. Otherwise you expose this class to a lot more functionality (everything storage related) than it actually needs. And then the `StorageProvider` will just expose all the boxes for other classes to use.

bisgardo replied:
Worth trying out but beware of overengineering :)

schwartz-concordium replied:
I did this refactoring and it came out quite nice. Kudos to you @orhoj 

It also solves one of the comments below. We have a single storage provider with the responsibility of opening connections and expose “boxes”.

The repository wrappers handle the queries and commands. In this case a `TermsAndConditionsRepository` wraps the read, write and update access to the `AcceptedTermsAndConditions`box.
These repository classes could live in network services and have relevant network filtering when accessing data. Since Terms and Condition are global it will live outside these.

Finally the `TermsAndConditionAcceptance` cubit holds the business logic.

I think it came out quite nice and resolved other comments given.

orhoj commented on test/providers/storage_test.dart:
```diff
@@ -0,0 +1,80 @@
+
+    // Assert
+    final actual = storage.getAcceptedTermsAndConditions(networkName);
+    expect(actual, isNotNull);
```
As `expectedVersion` is not null, this check appears to be redundant.

schwartz-concordium replied:
not sure what you mean. I’m validation `actual` isn’t null :-)

orhoj commented on lib/providers/storage.dart:
```diff
@@ -0,0 +1,44 @@
+  }
+
+  /// Opens all boxes asynchronously.
+  static Future<void> _openBoxes() async {
```
Opening all boxes means that we will always load all persisted data into memory. We generally have very little data so this might be fine, but the best practise would be to only open boxes on demand (and then close them again when done using them).

Have you considered this?

schwartz-concordium replied:
Yes, I initially considered using LazyBox to prevent everything from being fetched into memory, especially since we're dealing with relatively little data. 

However, upon further reflection after your comment, I agree that a combination of LazyBox and normal Box might be more appropriate. It's essential to differentiate not just based on the data volume but also on how frequently it needs to be loaded. For instance, in the case of the terms and conditions, which aren't accessed frequently, using a LazyBox to keep it out of memory makes perfect sense. I've already made the necessary changes to implement the use of LazyBox.

Additionally, incorporating metrics into our app would be highly beneficial. It would provide us with insights into potential bottlenecks and allow us to optimize performance effectively.

orhoj commented on lib/entities/accepted_terms_and_conditions.dart:
```diff
@@ -0,0 +1,26 @@
+class AcceptedTermsAndConditions {
+  static const table = "accepted_terms_and_conditions";
+
+  @HiveField(1)
```
Looking at the documentation `@HiveField` should be `0`-indexed, so start with `0`.

bisgardo committed:
Sync iOS project files

bisgardo committed:
Refactor: Split root widget

The current widget 'App' uses two nested `FutureBuilder`s in a wrong way: The provided future should not be constructed inside the build method. As the [documentation](https://api.flutter.dev/flutter/widgets/FutureBuilder-class.html#managing-the-future) puts it:

> The future must have been obtained earlier, e.g. during State.initState, State.didUpdateWidget, or State.didChangeDependencies. It must not be created during the State.build or StatelessWidget.build method call when constructing the FutureBuilder. If the future is created at the same time as the FutureBuilder, then every time the FutureBuilder's parent is rebuilt, the asynchronous task will be restarted.

This behavior has been observed during development as flickering and crashing due to duplicate network activation.

With this change the two `FutureBuilder`s now live in their own separate stateful widgets where the futures are constructed by their respective `initState` methods.

This change has the furtunate side effect of making the original widget much simpler and easier to understand.

schwartz-concordium commented on lib/providers/storage.dart:
```diff
@@ -0,0 +1,44 @@
+
+  /// Opens all boxes asynchronously.
+  static Future<void> _openBoxes() async {
+    final atcFuture = Hive.openBox<AcceptedTermsAndConditions>(AcceptedTermsAndConditions.table);
```
I made this class to show how boxes should be open asynchronously when more are added.
I will keep for now since we then avoid to refactor the structure.

Commit:
fix simple fixes

Commit:
comments

Commit:
Update terms_and_conditions.dart

bisgardo committed:
Declare fields as 'late final' instead of optional

Co-authored-by: Søren Schwartz <132270889+schwartz-concordium@users.noreply.github.com>

bisgardo committed:
Address review comments

Commit:
added terms and condition repository

Commit:
change to repository pattern

Commit:
format

Commit:
Merge branch 'mo/split-root-widget' into cbw-1474/migrate-to-hive

Commit:
refactor to lazy box

Commit:
format

Commit:
fix missing child and format

orhoj commented on lib/main.dart:
```diff
@@ -50,59 +50,145 @@ class App extends StatelessWidget {
+        future: _lastAccepted,
+        builder: (_, snapshot) {
+          // Check on the connection state since last accepted possible null
+          if (snapshot.connectionState != ConnectionState.done) {
```
So to me this is a bit of a code smell when we check directly on the `connectionState`. According to one of Flutter's examples (https://docs.flutter.dev/cookbook/networking/fetch-data#5-display-the-data) the main methods to use are:
```
snapshot.hasData
snapshot.hasError
```
I understand that `hasData` might be false when the Future is resolved as it might return `null`. However, assuming that it completed successfully when `snapshot.connectionState == ConnectionState.done` will also lead to errors, as the state is also `done` when there is an error.

So the current implementation handles the error case and the successful case in the same way, and I think that may be hard to understand and misleading. I think it would be nicer to use a default value (avoid `null` values) and rely on `hasData` and `hasError`.

The same issue is present for `_WithSelectedNetworkState`.

Commit:
Merge branch 'main' into cbw-1474/migrate-to-hive

Commit:
updated packages and changed type id

Commit:
updated with docs

Commit:
format

Commit:
added a future wrapper

orhoj commented on lib/main.dart:
```diff
@@ -158,6 +149,47 @@ class _WithSelectedNetworkState extends State<_WithSelectedNetwork> {
+    return FutureBuilder(
+        future: _lastAccepted,
+        builder: (_, snapshot) {
+          // Check on the connection state since last accepted possible null
```
This comment is no longer relevant. I also believe we should use the helper functions provided by the library and do it as:
```
if (snapshot.hasData) {
   return BlockProvider(... return TermsAndConditionAcceptance(_repository, snapshot.requireData.value));
} else if (snapshot.hasError) {
  // TODO Handle the error.
}
return const _Initializing();
```
As it is now, an error thrown by `_lastAccepted` will result in being stuck in `_Initializing`. So I only think it is safe to not make the `hasError` check if the function called cannot throw. And even then if it is changed in the future into something that throws, then you will have the same issue.

orhoj commented on lib/main.dart:
```diff
@@ -158,6 +149,47 @@ class _WithSelectedNetworkState extends State<_WithSelectedNetwork> {
+    final storage = context.read<ServiceRepository>().storage;
+    setState(() {
+      _repository = TermsAndConditionsRepository(storageProvider: storage);
+      _lastAccepted = _repository.getAcceptedTermsAndConditions().then((value) => FutureValue(value));
```
```suggestion
      _lastAccepted = _repository.getAcceptedTermsAndConditions().then(FutureValue.new);
```

Commit:
add todo for error

bisgardo commented:
@schwartz-concordium  I think it's unfortunate that this was merged without squashing. Please take a look at https://github.com/Concordium/concordium-wallet/commits/main and try to make sense of it. What commits belong to which PRs? It may be obvious in this case by looking at the author, but imagine that all PRs were like this. Then you'd have a commit history that's very far from being as clear as it could be, without any gains.

But also, many of the commits tell very little about why the changes were made. And many contain issues that were known and fixed before merging. This makes tools like git blame much less effective.

My strong personal view is that the commit history is the most useful when every commit is a progressive improvement (complete unit of work if you will) and gives a clear explanation of why it's done. This can certainly be done properly without squashing but it's much more work (i.e. rebasing) and is in my opinion done better just as multiple PRs.

I'm sorry for ranting but this matters a lot to me. I'll be very happy to do this discussion with you and the team offline.
